httpcore5/src/main/java/org/apache/hc/core5/http/impl/io/DefaultResponseOutOfOrderStrategy.java

```diff
@@ -0,0 +1,78 @@
+/*
+ * ====================================================================
+ * Licensed to the Apache Software Foundation (ASF) under one
+ * or more contributor license agreements.  See the NOTICE file
+ * distributed with this work for additional information
+ * regarding copyright ownership.  The ASF licenses this file
+ * to you under the Apache License, Version 2.0 (the
+ * "License"); you may not use this file except in compliance
+ * with the License.  You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ * ====================================================================
+ *
+ * This software consists of voluntary contributions made by many
+ * individuals on behalf of the Apache Software Foundation.  For more
+ * information on the Apache Software Foundation, please see
+ * <http://www.apache.org/>.
+ *
+ */
+
+package org.apache.hc.core5.http.impl.io;
+
+import org.apache.hc.core5.annotation.Contract;
+import org.apache.hc.core5.annotation.ThreadingBehavior;
+import org.apache.hc.core5.http.ClassicHttpRequest;
+import org.apache.hc.core5.http.io.HttpClientConnection;
+import org.apache.hc.core5.http.io.ResponseOutOfOrderStrategy;
+import org.apache.hc.core5.util.Timeout;
+
+import java.io.IOException;
+import java.io.InputStream;
+
+/**
+ * Default {@link ResponseOutOfOrderStrategy} implementation which checks for premature responses every 8 KiB based
+ * on testing using values between 4 KiB and 128 KiB. This is optimized for correctness and results in a maximum
+ * upload speed of 8 MiB/s.
+ *
+ * @since 5.1
+ */
+@Contract(threading = ThreadingBehavior.IMMUTABLE)
+public final class DefaultResponseOutOfOrderStrategy implements ResponseOutOfOrderStrategy {
+
+    public static final DefaultResponseOutOfOrderStrategy INSTANCE = new DefaultResponseOutOfOrderStrategy();
+
+    private static final int CHUNK_SIZE = 8 * 1024;
+
+    @Override
+    public boolean checkForEarlyResponse(
+            final HttpClientConnection connection,
+            final ClassicHttpRequest request,
+            final InputStream socketInputStream,
+            final long totalBytesSent,
+            final long nextWriteSize) throws IOException {
+        if (totalBytesSent == 0 || nextWriteBeginsNewChunk(totalBytesSent, nextWriteSize)) {
```

Comment:
Can `totalBytesSent == 0` actually happen? Does `totalBytes` not contain the length of the headers?

Reply:
That is correct, this retains the same behavior as the previous implementation, though ideally we would be able to avoid the check entirely for small requests. The API leaves the possibility open to pass the total length including headers, but I have not attempted to fix the behavior yet.

This caused a performance regression when I added `Content-Length: 0` using a zero-length ByteArrayEntity[1] because `write(empty-byte-array)` resulted in a 1ms response check despite sending no additional data.

1. Related to https://issues.apache.org/jira/browse/HTTPCORE-643

Reply:
I do remember that `totalBytes` does include headers bytes because I have seen the first buffer to contain headers + a bit of payload. Since you swapped `checkForEarlyResponse()` and the actual write, this will happen on the very first write -- at least as far as I understand the code. `nextWriteStartsNewChunk()` will also give a `false` for as long as the headers are smaller than the chunk size. Wouldn't that solve your problem for small requests? So the first check would happen past the chunk size. WDYT?

Reply:
I'm not sure I understand.

The order of the check and write is not impacted by this change as far as I can tell, before this PR the check was run prior to writes:

From 5.1.x: https://github.com/apache/httpcomponents-core/blob/6637591213221b21a5a473e7092fa2e65d69715c/httpcore5/src/main/java/org/apache/hc/core5/http/impl/io/DefaultBHttpClientConnection.java#L187-L188

Previously the first write would always incur a check because `chunks` was initialized at -1 rather than zero. A zero-length first write would result in `final long n = totalBytes / (8 * 1024);` being 0 and exceeding the `chunks` value of -1.
https://github.com/apache/httpcomponents-core/blob/6637591213221b21a5a473e7092fa2e65d69715c/httpcore5/src/main/java/org/apache/hc/core5/http/impl/io/DefaultBHttpClientConnection.java#L164-L175

I agree that the first packet/buffer is likely to contain both headers and request body content, but the OutputStream which counts bytes is only used for `sendRequestEntity` so I'm not sure how it can determine how many bytes were sent from `sendRequestHeader`.

Reply:
hmmm...right, we don't have the amount of bytes written to the output stream (@ok2c correct me if I am wrong). Though, I have definitively seen those bytes on the terminal during testing. Why not make an assumption that the vast amount of  headers will be much less than 1024 bytes and that first check will occur after the first chunk has been sent off? Its a tradeoff. A "default" implementation w/ specific assumptions. If someone is not happy, use a custom one or provide a flag for this too.

Reply:
I think that sounds reasonable, I imagine given the default 8k chunks there's going to be enough buffer space to send two chunks without a check. We may be able to simply remove the `totalBytes == 0` check.

Reply:
Exactly, this is my proposal. We skip the first check.

Reply:
Makes sense, I've added this optimization.